[srp-server] choose different UDP port from previous on reboot

superwhd opened the pull request:
This PR implements a mechanism to let an SRP server listen to a different UDP port on a reboot. 

- When an SRP server starts, it reads the UDP port used in the last time from `Settings`. Then the SRP server listens to the next port in the reserved range. 
- When SRP server successfully starts, store its UDP port in `Settings`.

abtink commented on src/core/net/srp_server.cpp:
```diff
@@ -438,15 +438,32 @@ void Server::CommitSrpUpdate(Error                    aError,
+    }
+#endif
 
+    VerifyOrExit(!IsRunning());
```
How about doing `VerifyOfExit(!IsRunning())` at the top of the method and before reading the info from `Settings`?

simonlingoogle replied:
Choose port after this line of code?

simonlingoogle commented on src/core/config/srp_server.h:
```diff
@@ -46,13 +46,37 @@
+#define OPENTHREAD_CONFIG_SRP_SERVER_UDP_PORT_MAX 53554
+#endif
+
+/**
+ * @def OPENTHREAD_CONFIG_SRP_SERVER_SAVE_INFO
+ *
+ * Define to 1 to enable SRP server feature to save its own port in non-volatile settings.
+ *
+ * When enabled, the SRP server will save the its port in the non-volatile settings. On a server
+ * restart (e.g., due to a device reset) it will choose a UDP port different from the saved one. The info is written to
+ * Settings after an SRP server successfully starts.
+ *
+ */
+#ifndef OPENTHREAD_CONFIG_SRP_SERVER_SAVE_INFO
+#define OPENTHREAD_CONFIG_SRP_SERVER_SAVE_INFO 1
```
```suggestion
```
always save info?

superwhd replied:
According to offline discussion, we decided to keep this macro.

simonlingoogle replied:
SGTM.

abtink reviewed (approved):
Looks great. Thanks @superwhd. 👍 
Few smaller suggestions below.

simonlingoogle commented on src/core/net/srp_server.cpp:
```diff
@@ -438,15 +438,32 @@ void Server::CommitSrpUpdate(Error                    aError,
+    uint16_t port  = kUdpPortMin;
 
-    VerifyOrExit(!IsRunning());
+#if OPENTHREAD_CONFIG_SRP_SERVER_ENABLE && OPENTHREAD_CONFIG_SRP_SERVER_SAVE_INFO
```
```suggestion
```

OPENTHREAD_CONFIG_SRP_SERVER_ENABLE must be 1 in this file?

simonlingoogle commented on src/core/net/srp_server.cpp:
```diff
@@ -438,15 +438,32 @@ void Server::CommitSrpUpdate(Error                    aError,
+    if (Get<Settings>().Read(info) == kErrorNone)
+    {
+        port = info.GetPort() + 1;
+        if (port > kUdpPortMax)
```
What if port < kUdpPortMin?

abtink replied:
👍  very good point @simonlingoogle. We need to handle such a case (i.e., the config min/max port range may be different between version that saved the value vs the one reading it later).

simonlingoogle commented on src/core/net/srp_server.cpp:
```diff
@@ -438,15 +438,32 @@ void Server::CommitSrpUpdate(Error                    aError,
 
     SuccessOrExit(error = PublishServerData());
 
+#if OPENTHREAD_CONFIG_SRP_SERVER_ENABLE && OPENTHREAD_CONFIG_SRP_SERVER_SAVE_INFO
+    info.SetPort(port);
+    IgnoreError(Get<Settings>().Save(info));
```
Save the port when the first SRP service is registered?

@librasungirl

simonlingoogle commented on src/core/net/srp_server.hpp:
```diff
@@ -78,7 +78,8 @@ class Server : public InstanceLocator, private NonCopyable
     {
-        kUdpPort = OPENTHREAD_CONFIG_SRP_SERVER_UDP_PORT, ///< The SRP Server UDP listening port.
+        kUdpPortMin = OPENTHREAD_CONFIG_SRP_SERVER_UDP_PORT_MIN, ///< The reserved min SRP Server UDP listening port.
+        kUdpPortMax = OPENTHREAD_CONFIG_SRP_SERVER_UDP_PORT_MAX, ///< The reserved max SRP Server UDP listening port.
```
`static_assert` that `kUdpPortMin <= kUdpPortMax` ?

simonlingoogle commented on tests/scripts/thread-cert/test_srp_server_reboot_port.py:
```diff
@@ -0,0 +1,125 @@
+#
+# Topology:
+#
+#   CLIENT (leader) -- SERVER (router)
```
```suggestion
#   CLIENT (Leader) -- SERVER (Router)
```

superwhd replied:
I switched to a style that most `test_srp*` files used.

simonlingoogle commented on tests/scripts/thread-cert/test_srp_server_reboot_port.py:
```diff
@@ -0,0 +1,125 @@
+    TOPOLOGY = {
+        CLIENT: {
+            'name': 'SRP_CLIENT',
+            'masterkey': '00112233445566778899aabbccddeeff',
```
```suggestion
```

simonlingoogle commented on tests/scripts/thread-cert/test_srp_server_reboot_port.py:
```diff
@@ -0,0 +1,125 @@
+        },
+        SERVER: {
+            'name': 'SRP_SERVER',
+            'masterkey': '00112233445566778899aabbccddeeff',
```
```suggestion
```

simonlingoogle commented on tests/scripts/thread-cert/test_srp_server_reboot_port.py:
```diff
@@ -0,0 +1,125 @@
+        ports = [server.get_srp_server_port()]
+
+        # Repeat the following steps for 20 times
+        for i in range(25):
```
We'd better define a constant in local file.

simonlingoogle commented on tests/scripts/thread-cert/test_srp_server_reboot_port.py:
```diff
@@ -0,0 +1,125 @@
+            self.simulator.go(5)
+            self.assertEqual(client.srp_client_get_state(), 'Enabled')
+            self.assertEqual(client.srp_client_get_server_address(), server.get_mleid())
+            self.assertNotEqual(old_port, server.get_srp_server_port())
```
I think we also need to verify that the SRP client will re-register its SRP service to the SRP server after SRP server is reset, which is the real purpose of this change.

simonlingoogle commented on src/core/config/srp_server.h:
```diff
@@ -46,13 +46,37 @@
+ *
+ */
+#ifndef OPENTHREAD_CONFIG_SRP_SERVER_SAVE_INFO
+#define OPENTHREAD_CONFIG_SRP_SERVER_SAVE_INFO 1
```
```suggestion
#define OPENTHREAD_CONFIG_SRP_SERVER_PORT_SWITCH_ENABLE 1
```
I think the configuration name should be what the feature is, rather than how we implemented it?

simonlingoogle commented on src/core/net/srp_server.cpp:
```diff
@@ -438,15 +438,32 @@ void Server::CommitSrpUpdate(Error                    aError,
     VerifyOrExit(!IsRunning());
 
+#if OPENTHREAD_CONFIG_SRP_SERVER_SAVE_INFO
+    Settings::SrpServerInfo info;
+    if (Get<Settings>().Read(info) == kErrorNone)
+    {
+        port = info.GetPort() + 1;
+        if (port < kUdpPortMin || port > kUdpPortMax)
+        {
+            port = kUdpPortMin;
+        }
+    }
```
```suggestion
{
    Settings::SrpServerInfo info;
    
    if (Get<Settings>().Read(info) == kErrorNone)
    {
        port = info.GetPort() + 1;
        if (port < kUdpPortMin || port > kUdpPortMax)
        {
            port = kUdpPortMin;
        }
    }
}
```
Variable must be at the front, so...

simonlingoogle commented on tests/scripts/thread-cert/test_srp_server_reboot_port.py:
```diff
@@ -0,0 +1,180 @@
+CLIENT = 1
+SERVER = 2
+
+REBOOT_TIMES = 25
```
```suggestion
REBOOT_TIMES = 3
```
maybe we don't need to reboot so many times?

superwhd replied:
I meant to let it go over all reserved ports and begin again from `kUdpPortMin`.

simonlingoogle reviewed (commented):
SGTM overall. 

But we need to decide when is the best time to save the port. 

@librasungirl has proposed to save the port when the first SRP client successfully registered a service, which I agree is the best strategy. 

Thoughts?

@superwhd @librasungirl @jwhui @wgtdkp @abtink

wgtdkp commented on include/openthread/platform/settings.h:
```diff
@@ -72,6 +72,7 @@ enum
     OT_SETTINGS_KEY_ON_LINK_PREFIX       = 0x000a, ///< On-link prefix for infrastructure link.
     OT_SETTINGS_KEY_SRP_ECDSA_KEY        = 0x000b, ///< SRP client ECDSA public/private key pair.
     OT_SETTINGS_KEY_SRP_CLIENT_INFO      = 0x000c, ///< The SRP client info (selected SRP server address).
+    OT_SETTINGS_KEY_SRP_SERVER_INFO      = 0x000d, ///< The SRP server info (address and port).
```
```suggestion
    OT_SETTINGS_KEY_SRP_SERVER_INFO      = 0x000d, ///< The SRP server info (UDP port).
```

superwhd replied:
Good catch! :+1:

wgtdkp reviewed (approved):
LGTM :+1: Thanks!

superwhd commented:
> SGTM overall.
> 
> But we need to decide when is the best time to save the port.
> 
> @librasungirl has proposed to save the port when the first SRP client successfully registered a service, which I agree is the best strategy.
> 
> Thoughts?
> 
> @superwhd @librasungirl @jwhui @wgtdkp @abtink

So far I'm still inclined to save the port when the SRP server starts. Reasons: 1) We have reserved 20 ports for this purpose so it's rare that all ports are used up in a short period of time. 2) It takes a bit more efforts to implement 'save the port when the first service registered' and the logic has to be added at two places, which makes it a bit harder to maintain.

wgtdkp commented:
> SGTM overall.
> 
> But we need to decide when is the best time to save the port.
> 
> @librasungirl has proposed to save the port when the first SRP client successfully registered a service, which I agree is the best strategy.
> 
> Thoughts?
> 
> @superwhd @librasungirl @jwhui @wgtdkp @abtink

I agree with @superwhd that there are not many benefits for saving only after the first service registration. Changing the port when there is no service doesn't introduce burdens since there is no SRP client. But saving port at start makes the code simpler and more reasonable.

simonlingoogle commented:
> > SGTM overall.
> > But we need to decide when is the best time to save the port.
> > @librasungirl has proposed to save the port when the first SRP client successfully registered a service, which I agree is the best strategy.
> > Thoughts?
> > @superwhd @librasungirl @jwhui @wgtdkp @abtink
> 
> I agree with @superwhd that there are not many benefits for saving only after the first service registration. Changing the port when there is no service doesn't introduce burdens since there is no SRP client. But saving port at start makes the code simpler and more reasonable.

We are comparing 2 strategies:
1. Current: Save port when the `SRP service TLV` is added to the local network data
2. Alternative: Save the port when a SRP service is registered by the first SRP client. 

Basically the `Alternative` strategy delays saving port until it has made sure that the `SRP service TLV` has been successfully added to the Leader's network data because some Thread device has been aware of the change. 

I think the `Alternative` solution has advantages:
1. When the reserved port range is small (e.g. 2 ports), the BR is less possible to use the same port after multiple fast reboots (SRP server will keep using the same port if there is no SRP registration). 
2. When there are SED/SSED children, the BR will not save the port until a SED child has registered a SRP service, which reduces the chance that multiple reboots may cause BR to switch back to the original port. 
3. After #6587 is merged, the SRP Server can no longer guarantee that its SRP service TLV will eventually be added the Network Data. For example, if there are 3 SRP servers, only 2 SRP servers will adds their SRP service TLV to the network data. The one that didn't add its SRP service TLV to the Network Data should better not change port.

Again, I don't see a significant difference between the two strategies. But the `Alternative` solution seems to be more robust in some corner cases. So, I believe we should implement it if it's not too complex.

Thoughts? @superwhd @wgtdkp @librasungirl

librasungirl commented on src/core/config/srp_server.h:
```diff
@@ -46,13 +46,37 @@
+#endif
+
+/**
+ * @def OPENTHREAD_CONFIG_SRP_SERVER_SAVE_INFO
```
```suggestion
 * @def OPENTHREAD_CONFIG_SRP_SERVER_PORT_SWITCH_ENABLE
```

librasungirl commented on src/core/config/srp_server.h:
```diff
@@ -46,13 +46,37 @@
+ *
+ * Define to 1 to enable SRP server feature to save its own port in non-volatile settings.
+ *
+ * When enabled, the SRP server will save the its port in the non-volatile settings. On a server
+ * restart (e.g., due to a device reset) it will choose a UDP port different from the saved one. The info is written to
+ * Settings after an SRP server successfully starts.
```
```suggestion
 * When enabled, the SRP server will save its port in the non-volatile settings. On a server
 * restart (e.g., due to a device reset) it will restore the port and change to a different one. The info is written to
 * Settings after the SRP server successfully starts.
```

librasungirl commented:
Agree with @simonlingoogle 's arguments. The optimized `Alternative` option may also be good to the lifetime of the flash as it only stores when necessary, reducing the write/erase operation.

Fine to me to have the optimization in separate PR.

librasungirl commented on src/core/common/settings.hpp:
```diff
@@ -677,6 +678,48 @@ class SettingsBase : public InstanceLocator
+
+        uint16_t mPort; // (in little-endian encoding)
+    } OT_TOOL_PACKED_END;
+#endif // OPENTHREAD_CONFIG_SRP_SERVER_ENABLE && OPENTHREAD_CONFIG_SRP_SERVER_SAVE_INFO
```
```suggestion
#endif // OPENTHREAD_CONFIG_SRP_SERVER_ENABLE && OPENTHREAD_CONFIG_SRP_SERVER_PORT_SWITCH_ENABLE
```

superwhd commented:
Changed the logic to save the port number when the first port is registered. Also added a test case to verify that the port number won't change if there's no service registered in the last run.

simonlingoogle commented on src/core/net/srp_server.hpp:
```diff
@@ -672,6 +674,7 @@ class Server : public InstanceLocator, private NonCopyable
 
     ServiceUpdateId mServiceUpdateId;
     bool            mEnabled;
+    bool            mHasRegisteredAnyService;
```
```suggestion
    bool            mEnabled: 1;
    bool            mHasRegisteredAnyService: 1;
```
nit

simonlingoogle commented on src/core/net/srp_server.cpp:
```diff
@@ -420,6 +421,15 @@ void Server::CommitSrpUpdate(Error                    aError,
         AddHost(&aHost);
+#if OPENTHREAD_CONFIG_SRP_SERVER_PORT_SWITCH_ENABLE
+        if (!mHasRegisteredAnyService)
+        {
+            mHasRegisteredAnyService = true;
+            Settings::SrpServerInfo info;
+            info.SetPort(mSocket.mSockName.mPort);
+            IgnoreError(Get<Settings>().Save(info));
+        }
```
```suggestion
        {
            Settings::SrpServerInfo info;

            mHasRegisteredAnyService = true;
            info.SetPort(mSocket.mSockName.mPort);
            IgnoreError(Get<Settings>().Save(info));
        }
```
style

simonlingoogle reviewed (approved):
LGTM :+1: 

with nits

librasungirl reviewed (approved):
👍

jwhui commented on src/core/config/srp_server.h:
```diff
@@ -46,13 +46,37 @@
+ *
+ * Define to 1 to enable SRP server feature to save its own port in non-volatile settings.
+ *
+ * When enabled, the SRP server will save its port in the non-volatile settings. On a server
+ * restart (e.g., due to a device reset) it will restore the port and change to a different one. The info is written to
+ * Settings after the SRP server successfully starts.
```
```suggestion
 * When enabled, the SRP server will save its port in the non-volatile settings. On a server
 * restart (e.g., due to a device reset) it will restore the port and change to a different one. The info is written to
 * Settings after the first service has been registered due to receiving an SRP Update.
```

jwhui reviewed (commented):
Looks great overall. One minor suggestion to update the doc to reflect the latest implementation.

abtink commented on src/core/net/srp_server.cpp:
```diff
@@ -438,13 +449,27 @@ void Server::CommitSrpUpdate(Error                    aError,
-    SuccessOrExit(error = mSocket.Open(HandleUdpReceive, this));
-    SuccessOrExit(error = mSocket.Bind(kUdpPort, OT_NETIF_THREAD));
+    {
+#if OPENTHREAD_CONFIG_SRP_SERVER_PORT_SWITCH_ENABLE
```
Move the braces inside the `#if` block?

abtink reviewed (approved):
LGTM. Thanks.
